beamer/health/check.py

```diff
@@ -389,6 +441,26 @@ def analyze_transfers(transfers: TransferMap, ctx: Context) -> None:
         else "Everything is calm in Beamerland."
     )
 
+    def format_token_volume(token_amount: int, token_details: TokenDetails) -> str:
+        return (
+            str(get_token_amount_in_decimals(token_amount, token_details))
+            + " "
+            + token_details["symbol"]
+        )
+
+    volume_per_token = ", ".join(
+        [
+            format_token_volume(token_amount, ctx.tokens[token_symbol])
+            for token_symbol, token_amount in ctx.stats.volume.items()
+        ]
+    )
```

Comment:
So, if i understood it right, after applying the changes like so:
```
expired_request_volume_per_token = ", ".join(
      format_token_volume(token_amount, ctx.tokens[token_symbol])
      for token_symbol, token_amount in ctx.stats.expired_requests_volume.items()
  )
```
we are basically passing a generator to the join function which will internally exhaust it to the point where it concatenates the strings with the delimiter ", " in between?
The difference is that instead of constructing the array from the generator we are passing the generator as argument?

Thanks for the awesome summary of the differences between the two and how they can be applied in different scenarios!

Reply:
> > So, if i understood it right, after applying the changes like so:
> 
> ```
> expired_request_volume_per_token = ", ".join(
>       format_token_volume(token_amount, ctx.tokens[token_symbol])
>       for token_symbol, token_amount in ctx.stats.expired_requests_volume.items()
>   )
> ```
> 
> we are basically passing a generator to the join function which will internally exhaust it to the point where it concatenates the strings with the delimiter ", " in between? The difference is that instead of constructing the array from the generator we are passing the generator as argument?

Yes, exactly. Pulling values out of the generator will be driven by `join`, and the generator itself will then ask for the next item of `ctx.stats.expired_requests_volume.items()`, transform it according to the expression you wrote and return it back to `join`.